[Memory Snapshot] Stop clearing history when changing context

aaronenyeshi opened the pull request:
Summary:
This change will avoid clearing the memory event history, when changing the context from `record_memory_history(context=None)` to `record_memory_history(context="python")`.

Now it will continue recording memory events with changing context on the fly. Only `record_memory_history(enabled=None)` will clear the history.

Test Plan:
# Ran on the following local Resnet50 example:

- At iteration=0, record_memory_history(context=None, stacks="python")
- At iteration=3, record_memory_history(context="all", stacks="python")
- After iteration=4, export_memory_snapshot()

## Before:
 - Only collects the last 2 iterations with python call stacks.
![image](https://github.com/pytorch/pytorch/assets/17602366/86154532-9f73-4d10-9194-19e8c96ee4f3)


## After:
 - Collects all 5 iterations, where first 3 iterations have no call stacks, and last 2 iterations have python call stacks.
![image](https://github.com/pytorch/pytorch/assets/17602366/c2c277d6-b400-4da2-85c8-a7f119d409f8)
![image](https://github.com/pytorch/pytorch/assets/17602366/dc9da2f8-41cc-44b0-9c32-ec3cbe79d2c4)


Differential Revision: D54084017

Pulled By: aaronenyeshi

albanD reviewed (commented):
Sounds good.
Should we have tests for this?

aaronenyeshi committed:
[Memory Snapshot] Stop clearing history when changing context (#120436)

Summary:

This change will avoid clearing the memory event history, when changing the context from `record_memory_history(context=None)` to `record_memory_history(context="python")`.

Now it will continue recording memory events with changing context on the fly. Only `record_memory_history(enabled=None)` will clear the history.

Test Plan: Added a unit test to test/test_cuda.py.

Reviewed By: leitian, tianfengfrank

Differential Revision: D54084017

Pulled By: aaronenyeshi

aaronenyeshi commented:
> Sounds good. Should we have tests for this?

@albanD , great point! I've added a new unit test to check this works as expected!

aaronenyeshi commented:
@pytorchbot merge
